File: examples/attitude/euler_angle_output.py
# /// script
# dependencies = ["brahe"]
# ///

"""
Demonstrates how to access and output Euler angle components.
"""

import brahe as bh
import numpy as np
import math

# Create Euler angles (ZYX: 45° yaw, 30° pitch, 15° roll)
ea = bh.EulerAngle(
    math.radians(45.0),
    math.radians(30.0),
    math.radians(15.0),
    bh.EulerAngleOrder.ZYX,
    bh.AngleFormat.RADIANS,
)

# Access individual angles
print("Individual angles (radians):")
print(f"  angle1 (Yaw/Z):   {ea.phi:.6f}")
print(f"  angle2 (Pitch/Y): {ea.theta:.6f}")
print(f"  angle3 (Roll/X):  {ea.psi:.6f}")

# Convert to degrees for readability
print("\nIndividual angles (degrees):")
print(f"  angle1 (Yaw/Z):   {math.degrees(ea.phi):.3f}°")
print(f"  angle2 (Pitch/Y): {math.degrees(ea.theta):.3f}°")
print(f"  angle3 (Roll/X):  {math.degrees(ea.psi):.3f}°")

# Access sequence order
print(f"\nSequence order: {ea.order}")

# Create vector from components [phi, theta, psi]
vec = np.array([ea.phi, ea.theta, ea.psi])
print("\nAs vector [phi, theta, psi] (radians):")
print(f"  {vec}")

# String representation
print("\nString representation:")
print(f"  {ea}")

# Expected output:
# Individual angles (radians):
#   angle1 (Yaw/Z):   0.785398
#   angle2 (Pitch/Y): 0.523599
#   angle3 (Roll/X):  0.261799
#
# Individual angles (degrees):
#   angle1 (Yaw/Z):   45.000°
#   angle2 (Pitch/Y): 30.000°
#   angle3 (Roll/X):  15.000°
#
# Sequence order: ZYX
#
# As vector [phi, theta, psi] (radians):
#   [0.78539816 0.52359878 0.26179939]
#
# String representation:
#   EulerAngle { angle1: 0.7853981633974483, angle2: 0.5235987755982988, angle3: 0.2617993877991494, order: ZYX }

File: examples/attitude/euler_angle_output.rs
//! Demonstrates how to access and output Euler angle components.

#[allow(unused_imports)]
use brahe as bh;
use brahe::attitude::FromAttitude;
use brahe::attitude::ToAttitude;

fn main() {
    // Create Euler angles (ZYX: 45° yaw, 30° pitch, 15° roll)
    let ea = bh::EulerAngle::new(
        bh::EulerAngleOrder::ZYX,
        (45.0_f64).to_radians(),
        (30.0_f64).to_radians(),
        (15.0_f64).to_radians(),
        bh::AngleFormat::Radians
    );

    // Access individual angles
    println!("Individual angles (radians):");
    println!("  angle1 (Yaw/Z):   {:.6}", ea.phi);
    println!("  angle2 (Pitch/Y): {:.6}", ea.theta);
    println!("  angle3 (Roll/X):  {:.6}", ea.psi);

    // Convert to degrees for readability
    println!("\nIndividual angles (degrees):");
    println!("  angle1 (Yaw/Z):   {:.3}°", ea.phi.to_degrees());
    println!("  angle2 (Pitch/Y): {:.3}°", ea.theta.to_degrees());
    println!("  angle3 (Roll/X):  {:.3}°", ea.psi.to_degrees());

    // Access sequence order
    println!("\nSequence order: {:?}", ea.order);

    // Manual vector representation [angle1, angle2, angle3]
    println!("\nAs vector [angle1, angle2, angle3] (radians):");
    println!("  [{:.6}, {:.6}, {:.6}]", ea.phi, ea.theta, ea.psi);

    // String representation
    println!("\nString representation:");
    println!("  {:?}", ea);
}

// Expected output:
// Individual angles (radians):
//   angle1 (Yaw/Z):   0.785398
//   angle2 (Pitch/Y): 0.523599
//   angle3 (Roll/X):  0.261799
//
// Individual angles (degrees):
//   angle1 (Yaw/Z):   45.000°
//   angle2 (Pitch/Y): 30.000°
//   angle3 (Roll/X):  15.000°
//
// Sequence order: ZYX
//
// As vector [angle1, angle2, angle3] (radians):
//   [0.7853981633974483, 0.5235987755982988, 0.2617993877991494]
//
// String representation:
//   EulerAngle { angle1: 0.7853981633974483, angle2: 0.5235987755982988, angle3: 0.2617993877991494, order: ZYX }
